### PR TITLE
final-style unzipping for stax2 prototype

### PR DESCRIPTION
**Pull request by jekbradbury**
Builds on Dougal and Matt's existing staxperiments ideas+code :slightly_smiling_face: 

The main difference is that it's based on a final-style unzip transformation rather than one that operates on jaxprs; in general I'd rather our NN things didn't have the same limitations as `jit` if we can avoid it. This is a choice that should be questioned though! (e.g. it means that initialization "does flops" even when it doesn't necessarily have to; maybe we could try abstracting to shape level and back off if there's an error?).

Also includes an internally-oriented `custom_transforms`-like-thing that differs from `custom_transforms` in a couple ways that may or may not have been good ideas:
- it's completely transparent to interpreters that it doesn't overload, which means it can wrap functions 
that use Python control flow
- it doesn't support pytrees, but instead follows the JAX-internal convention that arguments called positionally are dynamic arguments while those called keywordically are static parameters (note that this clashes with `jit`'s `static_argnums` convention, leading to additional ugliness in random.py)
- I can't imagine it does the right thing with closed-over tracers, so let's say it doesn't support them

### CODE REVIEW

**Comment by JuliusKunze** [comment]
@jekbradbury I woke up to an idea for
- refactoring JAXnet to use compositions of tagging transformations only and, relatedly,
- getting rid of explicit random key splitting/distribution for parameter initialization in the stax2 prototype, thereby dissolving the concern I expressed in https://github.com/google/jax/issues/1663#issuecomment-552803696.

The idea is to write a model as a combined init/apply function, but each parameter is initialized inline from an **individual (dummy) random key**. Both the random key and the parameter are tagged:

```python
def parameter(shape, init, name):
    key = lax.tag(PRNGKey(0), f'rng_{name}')
    return lax.tag(init(key, shape), name)


def Dense(out_dim, kernel_init=glorot(), bias_init=randn()):
    def dense(inputs):
        kernel = parameter((inputs.shape[-1], out_dim), kernel_init, 'kernel')
        bias = parameter((out_dim,), bias_init, 'bias')
        return np.dot(inputs, kernel) + bias

    return dense
```

This should allow to extract `init_parameters` and `apply` functions just using `collect`/`inject` from https://github.com/google/jax/pull/1660:

```python
@curry
def init_parameters_fun(fun, rng, *example_inputs):
    is_rng_name = lambda name: name.startswith('rng_')
    is_parameter_name = lambda name: not is_rng_name(name)

    def inject_all_rngs(fun, rng_names, rng):
        rng_dict = {}

        for name in rng_names:
            rng_, rng = random.split(rng)
            rng_dict[name] = rng_

        return inject(fun, rng_dict)

    _, tagged_dict = collect(fun)(*example_inputs)
    names = tagged_dict.keys()
    rng_names = list(filter(is_rng_name, names))
    parameter_names = list(filter(is_parameter_name, names))

    rng_injected_fun = inject_all_rngs(fun, rng_names, rng)
    _, tagged_dict = collect(rng_injected_fun)(rng, *example_inputs)
    return {tagged_dict[p] for p in parameter_names}


@curry
def apply_fun(fun, parameters, *example_inputs):
    return inject(fun, parameters)(*example_inputs)


dense = Dense(3)
parameters = init_parameters_fun(dense)(PRNGKey(0), np.zeros(1, 3))
outputs = apply_fun(dense)(parameters, np.zeros(1, 3))
```

I don't understand your explanation for `lax.push_tag` / metadata stacks / namespaces in https://github.com/google/jax/pull/1660 well enough to predict how much this helps for supporting nested parametrized functions / parameter structure (as currently implemented in JAXnet). Could you provide an example? I tried to understand it via your tests, but I am confused about [`test_collect_nested`](https://github.com/google/jax/pull/1660/files#diff-06ab2b0737d4f2ec271cadcee9ac2ff7R59) / [`test_inject_nested`](https://github.com/google/jax/pull/1660/files#diff-06ab2b0737d4f2ec271cadcee9ac2ff7R98): How can the string `"y"` be used as a JAX value?

Remarkably, this seems to require two `collect` calls in `init_parameters_fun`, with the first one parsing the occurring random key names that are to be injected. I wonder whether there is a way to combine this call with the following `inject` call, i. e. providing a version of `inject` that injects into a subset of all tags where the name fulfills a specified property, and the injected values are specified via a callback.

**Comment by gehring** [comment]
@JuliusKunze The heavy dependence on naming is giving me `tf.variable_scope` flashbacks. Any thoughts on how the same pitfalls that made variable scopes clunky can be avoided here? Hopefully this wouldn't require the same kind of heavy machinery used by Sonnet v1 for scope tracking.

**Comment by JuliusKunze** [comment]
@gehring Very much agree, a horrible aspect of TF1 variable management was that parameter sharing was controlled via names and the resulting horribly intransparent cross-dependencies. Under the assumption that tags with the same (path) name will cause an error, this will not be an issue. An interesting question is then how parameter sharing could be implemented using the tagging transformations! It's unclear to me how this would work without nested tracing / object identity ([like in JAXnet](https://github.com/JuliusKunze/jaxnet/blob/master/API.md#parameter-sharing)).

The `startswith('rng_')` part is obviously super unreliable (just used it to showcase the idea), but this could potentially be made robust with more advanced metadata infrastructure announced in https://github.com/google/jax/pull/1660 (waiting for more details here).

**Comment by jekbradbury** [comment]
> I don't understand your explanation for lax.push_tag / metadata stacks / namespaces in #1660 well enough to predict how much this helps for supporting nested parametrized functions / parameter structure (as currently implemented in JAXnet). Could you provide an example?

First, an example which conflates RNG keys and parameter scopes (but can still express weight sharing):

```python
@curry
def Dense(features_out, key, x):
  features_in = x.shape[-1]
  W = initializers.normal()(random.fold_in(key, "W"), (features_in, features_out))
  b = initializers.normal()(random.fold_in(key, "b"), (features_out,))
  return np.dot(x, W) + b

@curry
def Serial(layers, x):
  for layer in layers:
    x = layer(x)
  return x

def my_net(key, x):
  layer_1 = Dense(2, random.fold_in(key, "dense_1"))
  layer_2 = Dense(x.shape[0], random.fold_in(key, "dense_2"))
  net = Serial([layer_1, layer_2, layer_1])
  return net(x)

x = np.ones(3)
init_key = random.PRNGKey(0)
params = collect(my_net)(init_key, x)
apply_key = random.PRNGKey(0)
y = inject(my_net, params)(apply_key, x)
```

This is cute, but there's one slightly magical thing going on: the argument `key` is playing multiple roles. At initialization time, it's carrying the value of the initialization key. At application time, it's carrying the value of a runtime PRNG key (e.g. for dropout; unused here). And at both times, it's carrying scope/tag metadata through its data dependency.

You can also split out the RNG and scope roles:

```python
@curry
def Dense(features_out, key, scope, x):
  features_in = x.shape[-1]
  W_key, b_key = random.split(key)
  W = lax.tag(random.normal(W_key, (features_in, features_out)), "W", scope)
  b = lax.tag(random.normal(b_key, (features_out,)), "b", scope)
  return np.dot(x, W) + b

@curry
def Serial(layers, x):
  for layer in layers:
    x = layer(x)
  return x

def my_net(key, scope, x):
  key_1, key_2 = random.split(key)
  layer_1 = Dense(2, key_1, lax.push_tag(scope, "dense_1"))
  layer_2 = Dense(x.shape[0], key_2, lax.push_tag(scope, "dense_2"))
  net = Serial([layer_1, layer_2, layer_1])
  return net(x)

x = np.ones(3)
init_key = random.PRNGKey(0)
scope = 0. # can be any JAX value at all; we only use it for metadata
params = collect(my_net)(init_key, scope, x)
apply_key = random.PRNGKey(0)
y = inject(my_net, params)(apply_key, scope, x)
```

But you can see why it might be convenient to use the same argument for both (it's related to the question in a framework like Sonnet of whether you want to maintain a single "module stack" or separate stacks for parameters, submodules, etc.).

Is that helpful? Note that in this scheme parameter sharing _is_ driven by names, but (one way to see it is that) it's driven specifically by names folded into RNG keys, and those names are going to be mostly function-local (because of the hierarchy). While there are lots of benefits of using Python object identity to drive parameter sharing instead, it's a tricky choice in JAX because tracers can never depend on Python object identity behavior—it's completely lost under `jit`! This is essentially because depending on Python object identity means breaking [referential transparency](https://en.wikipedia.org/wiki/Referential_transparency), since two modules can have the same _value_ but may or may not have the same identity.

**Comment by JuliusKunze** [comment]
@jekbradbury Very helpful, thank you.

What's the behavior of
```python
test = lambda: lax.tag(1, "a") + lax.tag(2, "a")
print(collect(test))
```

and `print(inject(test, {"a": 1}))`?

**Comment by jekbradbury** [comment]
That function has no data dependence on inputs, so none of the JAX tracers will see what happens in it. We could create a dynamically scoped tracer instead, but its interaction with other transformations would be very hard, maybe impossible, to get right.

**Comment by JuliusKunze** [comment]
What about

```
test = lambda x: lax.tag(x + 1, "a") + lax.tag(x + 2, "a")
print(collect(test)(0))
print(inject(test, {"a": 1})(0))
```

**Comment by jekbradbury** [comment]
I'm not sure what that should do. I can see arguments for an error (the current behavior) or for the second tagged value to be replaced by the value yielded in the first one (this might make it easier to build libraries like Sonnet/jaxnet on top of tagging).

But I'm leaning towards an error because of examples like this one: here, the two values don't have a data dependency on each other, so the mutual order in which they're tagged isn't well-defined (due to toposorting and our assumption of no side effects, `partial_eval` might reorder such independent primitive operations relative to the order they were executed in the Python interpreter).

**Comment by JuliusKunze** [comment]
@jekbradbury +1 for a loud error. Parameter sharing should not require specifying multiple redundant (and therefore possibly inconsistent) initialization functions in the first place.

Will any object be usable as metadata for `lax.push_tag` (i. e. not just strings)? 

Regarding: "While there are lots of benefits of using Python object identity to drive parameter sharing instead, it's a tricky choice in JAX because tracers can never depend on Python object identity behavior—it's completely lost under jit!" JAXnet allows both by using object identity of ([`parametrized`](https://github.com/JuliusKunze/jaxnet/blob/master/jaxnet/core.py#L23)) _primitives_, which is [not lost](https://github.com/JuliusKunze/jaxnet/blob/master/tests/test_core.py#L199) under transformations like `jit`.
